File: src/net/opgenorth/yeg/buildings/data/SQLiteBuildingDataService.java
package net.opgenorth.yeg.buildings.data;

import android.content.ContextWrapper;
import android.database.Cursor;
import net.opgenorth.yeg.buildings.model.Building;
import net.opgenorth.yeg.buildings.model.IBuildingSorter;

import java.util.ArrayList;
import java.util.List;
import java.util.UUID;


public class SQLiteBuildingDataService implements IBuildingDataService {
    private ContextWrapper _context;

    public SQLiteBuildingDataService(ContextWrapper context) {
        this._context = context;
    }

    @Override
    public List<Building> fetchAll() {
        ArrayList<Building> list = new ArrayList<Building>();
        Cursor c = _context.getContentResolver().query(BuildingsContentProvider.Columns.CONTENT_URI, BuildingsContentProvider.Columns.ALL_COLUMNS, null, null, null);

        try {
            c.moveToFirst();
            do {
                Building building = getBuildingFromRow(c);
                list.add(building);
            } while (c.moveToNext());

        } finally {
            c.close();
        }

        return list;
    }

    @Override
    public List<Building> fetchAll(IBuildingSorter sortedBy) {
        return fetchAll();
    }

    @Override
    public boolean hasRecords() {
        // TODO:  there has to be a more efficient way to to this.
        boolean hasRecords = false;
        Cursor c = _context.getContentResolver().query(BuildingsContentProvider.Columns.CONTENT_URI, BuildingsContentProvider.Columns.ALL_COLUMNS, null, null, null);

        try {
            if (c.moveToFirst()) {
                long count = 0;
                do {
                    count++;
                }
                while (c.moveToNext());
                hasRecords = count > 0;
            }
        } catch (Exception ex) {
            hasRecords = false;
        } finally {
            c.close();
        }

        return hasRecords;
    }

    private Building getBuildingFromRow(Cursor c) {
        Building building = new Building();
        building.setId(c.getLong(0));
        building.setName(c.getString(1));
        building.setRowKey(UUID.fromString(c.getString(2)));
        building.setAddress(c.getString(3));
        building.setNeighbourHood(c.getString(4));
        building.setUrl(c.getString(5));
        building.setConstructionDate(c.getString(6));

        String latitude = c.getString(8);
        String longitude = c.getString(7);

        building.setLocation(Double.parseDouble(latitude), Double.parseDouble(longitude));
        return building;
    }
}

File: dotnet/src/yegbuildings/Data/SQLiteBuildingDataService.cs
using System;
using System.Collections.Generic;
using Android.Content;
using Android.Database;
using Android.Util;
using Java.Lang;
using Net.Opgenorth.Yeg.Buildings.Model;
using Exception = System.Exception;

namespace Net.Opgenorth.Yeg.Buildings.Data
{
    public class SQLiteBuildingDataService : IBuildingDataService
    {
        private readonly ContextWrapper _context;

        public SQLiteBuildingDataService(ContextWrapper context)
        {
            _context = context;
        }

        public bool HasRecords
        {
            get
            {
                bool hasRecords;
                long count = 0;
                var c = _context.ContentResolver.Query(Columns.CONTENT_URI, Columns.ALL_COLUMNS, null, null, null);
                try
                {
                    c.MoveToFirst();
                    do
                    {
                        count++;
                    }
                    while (c.MoveToNext());
                    hasRecords = count > 0;
                }
                catch (Exception ex)
                {
                    hasRecords = false;
                    Log.Debug(Constants.LOG_TAG, (Throwable) ex, "Problem to see if we have records.");
                }
                finally
                {
                    c.Close();
                }
                return hasRecords;
            }
        }

        private List<Building> FetchAllInternal()
        {
            var buildings = new List<Building>();
            var c = _context.ContentResolver.Query(Columns.CONTENT_URI, Columns.ALL_COLUMNS, null, null, null);
            try
            {
                c.MoveToFirst();
                do
                {
                    var building = GetBuildingFromRow(c);
                    buildings.Add(building);
                }
                while (c.MoveToNext());
            }
            finally
            {
                c.Close();
            }
            return buildings;
        }

        private static Building GetBuildingFromRow(ICursor cursor)
        {
            var building = new Building
                               {
                                   Id = cursor.GetLong(0),
                                   Name = cursor.GetString(1),
                                   RowKey = Guid.Parse(cursor.GetString(2)),
                                   Address = cursor.GetString(3),
                                   NeighbourHood = cursor.GetString(4),
                                   Url = cursor.GetString(5),
                                   ConstructionDate = cursor.GetString(6),
                                   Longitude = cursor.GetDouble(7),
                                   Latitude = cursor.GetDouble(8)
                               };
            return building;
        }

        public IList<Building> FetchAll()
        {
            return FetchAllInternal();
        }

        public IList<Building> FetchAll(IBuildingSorter sortedBy)
        {
            var buildings = FetchAllInternal();
            return sortedBy == null ? buildings : sortedBy.Sort(buildings);
        }
    }
}